File: Microservices-Backend/theater-service/src/main/java/com/revticket/theater/service/TheaterService.java
package com.revticket.theater.service;

import com.revticket.theater.dto.TheaterRequest;
import com.revticket.theater.dto.TheaterResponse;
import com.revticket.theater.dto.TheaterStatsResponse;
import com.revticket.theater.entity.Screen;
import com.revticket.theater.entity.Theater;
import com.revticket.theater.repository.ScreenRepository;
import com.revticket.theater.repository.TheaterRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
public class TheaterService {

    @Autowired
    private TheaterRepository theaterRepository;

    @Autowired
    private ScreenRepository screenRepository;

    @Transactional(readOnly = true)
    public List<TheaterResponse> getAllTheaters(boolean activeOnly) {
        List<Theater> theaters = activeOnly
                ? theaterRepository.findByIsActiveTrue()
                : theaterRepository.findAll();
        return theaters.stream()
                .map(this::mapToResponse)
                .collect(Collectors.toList());
    }

    @Transactional(readOnly = true)
    public List<TheaterResponse> getTheatersByCity(String city, boolean activeOnly) {
        List<Theater> theaters = activeOnly
                ? theaterRepository.findByLocationAndIsActiveTrue(city)
                : theaterRepository.findByLocation(city);
        return theaters.stream()
                .map(this::mapToResponse)
                .collect(Collectors.toList());
    }

    @Transactional(readOnly = true)
    public Optional<TheaterResponse> getTheaterById(String id) {
        return theaterRepository.findById(Objects.requireNonNullElse(id, "")).map(this::mapToResponse);
    }

    @Transactional
    public TheaterResponse createTheater(TheaterRequest request) {
        Theater theater = new Theater();
        applyRequest(theater, request);
        return mapToResponse(theaterRepository.save(theater));
    }

    @Transactional
    public TheaterResponse updateTheater(String id, TheaterRequest request) {
        Theater theater = theaterRepository.findById(Objects.requireNonNullElse(id, ""))
                .orElseThrow(() -> new RuntimeException("Theater not found"));
        applyRequest(theater, request);
        return mapToResponse(theaterRepository.save(theater));
    }

    @Transactional
    public TheaterResponse updateTheaterStatus(String id, boolean isActive) {
        Theater theater = theaterRepository.findById(Objects.requireNonNullElse(id, ""))
                .orElseThrow(() -> new RuntimeException("Theater not found"));
        theater.setIsActive(isActive);
        return mapToResponse(theaterRepository.save(theater));
    }

    @Transactional
    public void deleteTheater(String id) {
        Theater theater = theaterRepository.findById(Objects.requireNonNullElse(id, ""))
                .orElseThrow(() -> new RuntimeException("Theater not found"));
        theaterRepository.delete(theater);
    }

    private void applyRequest(Theater theater, TheaterRequest request) {
        theater.setName(request.getName());
        theater.setLocation(request.getLocation());
        theater.setAddress(request.getAddress());
        theater.setTotalScreens(request.getTotalScreens());
        theater.setImageUrl(request.getImageUrl());
        if (request.getIsActive() != null) {
            theater.setIsActive(request.getIsActive());
        } else if (theater.getIsActive() == null) {
            theater.setIsActive(true);
        }
    }

    public TheaterStatsResponse getTheaterStats() {
        Long totalTheaters = theaterRepository.count();
        Long totalScreens = screenRepository.count();
        Long totalSeats = screenRepository.findAll().stream()
                .mapToLong(screen -> screen.getTotalSeats() != null ? screen.getTotalSeats() : 0)
                .sum();
        return new TheaterStatsResponse(totalTheaters, totalScreens, totalSeats);
    }

    private TheaterResponse mapToResponse(Theater theater) {
        return TheaterResponse.builder()
                .id(theater.getId())
                .name(theater.getName())
                .location(theater.getLocation())
                .address(theater.getAddress())
                .totalScreens(theater.getTotalScreens())
                .imageUrl(theater.getImageUrl())
                .isActive(theater.getIsActive())
                .defaultCategories(null)
                .build();
    }
}

File: Frontend/src/app/core/services/theater.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';

export interface Theater {
  id: string;
  name: string;
  location: string;
  address: string;
  totalScreens?: number;
  imageUrl?: string;
  isActive: boolean;
}

export interface Screen {
  id: string;
  name: string;
  totalSeats: number;
  theaterId: string;
}

@Injectable({
  providedIn: 'root'
})
export class TheaterService {
  private http = inject(HttpClient);

  getAllTheaters(activeOnly: boolean = true, city?: string): Observable<Theater[]> {
    let params = new HttpParams();
    if (!activeOnly) {
      params = params.set('activeOnly', 'false');
    }
    if (city) {
      params = params.set('city', city);
    }
    return this.http.get<Theater[]>(`${environment.apiUrl}/theaters`, { params });
  }

  getTheaterById(id: string): Observable<Theater> {
    return this.http.get<Theater>(`${environment.apiUrl}/theaters/${id}`);
  }

  createTheater(theater: Partial<Theater>): Observable<Theater> {
    return this.http.post<Theater>(`${environment.apiUrl}/theaters`, theater);
  }

  updateTheater(id: string, theater: Partial<Theater>): Observable<Theater> {
    return this.http.put<Theater>(`${environment.apiUrl}/theaters/${id}`, theater);
  }

  updateTheaterStatus(id: string, isActive: boolean): Observable<Theater> {
    const params = new HttpParams().set('active', String(isActive));
    return this.http.patch<Theater>(`${environment.apiUrl}/theaters/${id}/status`, null, { params });
  }

  deleteTheater(id: string): Observable<void> {
    return this.http.delete<void>(`${environment.apiUrl}/theaters/${id}`);
  }

  getTheaterScreens(theaterId: string): Observable<Screen[]> {
    return this.http.get<Screen[]>(`${environment.apiUrl}/admin/theaters/${theaterId}/screens`);
  }

  getAdminTheaters(activeOnly: boolean = false): Observable<Theater[]> {
    let params = new HttpParams();
    if (activeOnly) {
      params = params.set('activeOnly', 'true');
    }
    return this.http.get<Theater[]>(`${environment.apiUrl}/theaters`, { params });
  }
}

